packages/zip-it-and-ship-it/src/rate_limit.ts

```diff
@@ -1,30 +1,69 @@
-export enum RateLimitAlgorithm {
-  SlidingWindow = 'sliding_window',
-}
+import { z } from 'zod'
 
-export enum RateLimitAggregator {
-  Domain = 'domain',
-  IP = 'ip',
+export interface TrafficRules {
+  action: {
+    type: string
+    config: {
+      rateLimitConfig: {
+        algorithm: string
+        windowSize: number
+        windowLimit: number
+      }
+      aggregate: {
+        keys: {
+          type: string
+        }[]
+      }
+      to?: string
+    }
+  }
 }
 
-export enum RateLimitAction {
-  Limit = 'rate_limit',
-  Rewrite = 'rewrite',
-}
+const rateLimitAction = z.enum(['rate_limit', 'rewrite'])
+const rateLimitAlgorithm = z.enum(['sliding_window'])
+const rateLimitAggregator = z.enum(['domain', 'ip'])
+const slidingWindow = z.object({
+  windowLimit: z.number(),
+  windowSize: z.number(),
+})
+const rewriteActionConfig = z.object({
+  to: z.string(),
+})
 
-interface SlidingWindow {
-  windowLimit: number
-  windowSize: number
-}
+export const rateLimit = z
+  .object({
+    action: rateLimitAction.optional(),
+    aggregateBy: rateLimitAggregator.or(z.array(rateLimitAggregator)).optional(),
+    algorithm: rateLimitAlgorithm.optional(),
+  })
+  .merge(slidingWindow)
```

Comment:
The way I understand it, both `slidingWindow` and `rewriteActionConfig` are very tightly related to `rateLimit`. Why compose them from three schemas, when you can also define them all in the same `.object`? Is it because there's subsets that are mutually exclusive or mutually required?

Reply:
I'm just going by the current implementation, but I think we can revisit this.